### PR TITLE
Fix all compilation and Code Analysis warnings

### LINKED ISSUE #2737: Fix all compilation warnings
Currently we get 5 compilation warnings and 97 code analysis warnings

### PR DESCRIPTION
**Pull request by mnaoumov**
Fixed #2737

### CODE REVIEW

File: GitCommands/Statistics/ImpactLoader.cs [hunk 1]
```diff
@@ -62,12 +70,19 @@ public ImpactLoader(IGitModule aModule)
...
         ~ImpactLoader()
         {
-            Stop();
+            Dispose(false);
         }
 
         public void Dispose()
         {
+            Dispose(true);
+        }
```
Comment (IDisposable) [inline GitCommands/Statistics/ImpactLoader.cs]: Shouldn't there be a GC.SuppressFinalize(this);

**Comment by mnaoumov** [comment]
Thanks @IDisposable , I missed that part in that file.

BTW, it's funny that this was found by you because of your nick name ;)

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 2]
```diff
@@ -217,7 +219,7 @@ public FormBrowse(GitUICommands aCommands, string filter)
             RevisionGrid.MenuCommands.MenuChanged += (sender, e) => _formBrowseMenus.OnMenuCommandsPropertyChanged();
         }
 
-        private void Translate()
+        private new void Translate()
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormBrowse.cs]: :bulb: I would prefer to see this renamed.

File: GitCommands/AsyncLoader.cs [hunk 3]
```diff
@@ -126,6 +126,17 @@ private bool OnLoadingError(Exception exception)
...
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_cancelledTokenSource != null)
+                _cancelledTokenSource.Dispose();
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
```
Comment (sharwell) [inline GitCommands/AsyncLoader.cs]: :exclamation: This should have the following form:

```
public void Dispose()
{
  Dispose(true);
  GC.SuppressFinalize(this);
}
```

File: GitCommands/AsyncLoader.cs [hunk 4]
```diff
@@ -126,6 +126,17 @@ private bool OnLoadingError(Exception exception)
             LoadingError(this, args);
             return args.Handled;
         }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_cancelledTokenSource != null)
```
Comment (sharwell) [inline GitCommands/AsyncLoader.cs]: :exclamation: You also need to check the previous value of `_cancellationTokenSource` at the point it is assigned in `Load()` above, and call `Dispose()` before discarding the old reference.

File: GitCommands/Logging/CommandLogger.cs [hunk 5]
```diff
@@ -6,11 +6,10 @@ namespace GitCommands.Logging
 {
     public sealed class CommandLogger
     {
-        public delegate void CommandsChangedHandler();
         private const int LogLimit = 500;
         private readonly Queue<CommandLogEntry> _logQueue = new Queue<CommandLogEntry>(LogLimit);
 
-        public event CommandsChangedHandler CommandsChanged = delegate { };
+        public event EventHandler CommandsChanged = delegate { };
```
Comment (sharwell) [inline GitCommands/Logging/CommandLogger.cs]: :bulb: I understand why it's assigning a default value of `delegate { }` (so you never need to check for null at the invocation site), but I don't like it because it introduces performance overhead for little (if any) gain in clarity.

File: GitCommands/Statistics/ImpactLoader.cs [hunk 6]
```diff
@@ -62,12 +70,20 @@ public ImpactLoader(IGitModule aModule)
 
         ~ImpactLoader()
```
Comment (sharwell) [inline GitCommands/Statistics/ImpactLoader.cs]: :exclamation: This finalizer is unnecessary and should be removed.

File: GitCommands/Statistics/ImpactLoader.cs [hunk 7]
```diff
@@ -62,12 +70,20 @@ public ImpactLoader(IGitModule aModule)
...
 
         public void Dispose()
         {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void Dispose(bool disposing)
+        {
             Stop();
```
Comment (sharwell) [inline GitCommands/Statistics/ImpactLoader.cs]: :exclamation: The call to `Stop` operates on managed resources, and therefore should only occur if `disposing` is true.

File: GitCommands/RevisionGraph.cs [hunk 8]
```diff
@@ -98,12 +98,21 @@ public RevisionGraph(GitModule module)
 
         ~RevisionGraph()
```
Comment (sharwell) [inline GitCommands/RevisionGraph.cs]: :exclamation: This finalizer is unnecessary and should be removed.

File: GitCommands/RevisionGraph.cs [hunk 9]
```diff
@@ -98,12 +98,21 @@ public RevisionGraph(GitModule module)
...
 
         public void Dispose()
         {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void Dispose(bool disposing)
+        {
             _backgroundLoader.Cancel();
```
Comment (sharwell) [inline GitCommands/RevisionGraph.cs]: `_backgroundLoader` is a managed resource, and can only be used if `disposing` is true.

File: GitCommands/Settings/SettingsCache.cs [hunk 10]
```diff
@@ -16,17 +16,17 @@ public SettingsCache()
...
+            Dispose(true);
             GC.SuppressFinalize(this);
         }
 
-        protected virtual void DisposeImpl()
+        protected virtual void Dispose(bool disposing)
         {
         }
 
         ~SettingsCache()
```
Comment (sharwell) [inline GitCommands/Settings/SettingsCache.cs]: :exclamation: This finalizer is unnecessary and should be removed.

File: GitCommands/Settings/FileSettingsCache.cs [hunk 11]
```diff
@@ -51,7 +51,9 @@ void _fileWatcher_Renamed(object sender, RenamedEventArgs e)
             FileChanged();
         }
 
-        protected override void DisposeImpl()
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2213:DisposableFieldsShouldBeDisposed", MessageId = "SaveTimer")]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2213:DisposableFieldsShouldBeDisposed", MessageId = "_fileWatcher")]
+        protected override void Dispose(bool disposing)
         {
             LockedAction(() =>
```
Comment (sharwell) [inline GitCommands/Settings/FileSettingsCache.cs]: :exclamation: `LockedAction` can only be called when `disposing` is true.

File: GitCommands/Utils/WeakRefCache.cs [hunk 12]
```diff
@@ -64,5 +64,22 @@ private void OnClearTimer(object source, System.Timers.ElapsedEventArgs e)
...
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+                _clearTimer.Dispose();
+        }
+
+        ~WeakRefCache()
```
Comment (sharwell) [inline GitCommands/Utils/WeakRefCache.cs]: :exclamation: This finalizer is unnecessary and should be removed.

File: GitUI/BuildServerIntegration/BuildServerWatcher.cs [hunk 13]
```diff
@@ -93,6 +93,7 @@ public void CancelBuildStatusFetchOperation()
             }
         }
 
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2202:Do not dispose objects multiple times")]
```
Comment (sharwell) [inline GitUI/BuildServerIntegration/BuildServerWatcher.cs]: :exclamation: This warning was not incorrect. The `MemoryStream` instance is disposed twice because the [`StreamWriter` constructor](https://msdn.microsoft.com/en-us/library/3aadshsx.aspx) takes ownership of the `Stream` it is created with.

File: GitUI/CommandsDialogs/RepoHosting/CreatePullRequestForm.cs [hunk 14]
```diff
@@ -191,5 +191,24 @@ private void _createBtn_Click(object sender, EventArgs e)
...
+        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _remoteLoader.Cancel();
+                _remoteLoader.Dispose();
+            }
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/CommandsDialogs/RepoHosting/CreatePullRequestForm.cs]: :bulb: this block could be added to the previous block to simplify the condition.

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/Dashboard.cs [hunk 15]
```diff
@@ -170,12 +171,12 @@ public void SaveSplitterPositions()
             }
         }
 
-        public event GitModuleChangedEventHandler GitModuleChanged;
+        public event EventHandler<GitModuleEventArgs> GitModuleChanged;
 
-        public virtual void OnModuleChanged(GitModule aModule)
+        public virtual void OnModuleChanged(object sender, GitModuleEventArgs e)
         {
             if (GitModuleChanged != null)
```
Comment (sharwell) [inline GitUI/CommandsDialogs/BrowseDialog/DashboardControl/Dashboard.cs]: This should be assigned to a temporary variable, to ensure it is not affected by code operating on other threads.

```csharp
var t = GitModuleChanged;
if (t != null)
  t(this, e);
```

In C# 6, this is simplified dramatically to this:

```csharp
GitModuleChanged?.Invoke(this, e);
```

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardItem.cs [hunk 16]
```diff
@@ -164,5 +164,23 @@ private void OnKeyDown(object sender, PreviewKeyDownEventArgs e)
...
                 OnClick(e);
         }
+
+        /// <summary> 
+        /// Clean up any resources being used.
+        /// </summary>
+        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
+        protected override void Dispose(bool disposing)
+        {
+            CancelBranchNameLoad();
```
Comment (sharwell) [inline GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardItem.cs]: :exclamation: This call should only occur if `disposing` is true.

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardItem.cs [hunk 17]
```diff
@@ -164,5 +164,23 @@ private void OnKeyDown(object sender, PreviewKeyDownEventArgs e)
...
+
+            if (disposing)
+                _branchNameLoader.Dispose();
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardItem.cs]: :bulb: This could be combined with the previous check for `disposing` to simplify the code.

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs [hunk 18]
```diff
@@ -179,7 +187,7 @@ public void ShowAuthor(bool show)
             this.DateColumn.Visible = show;
         }
 
-        [DefaultValue(FilterType.None)]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2002:DoNotLockOnObjectsWithWeakIdentity"), DefaultValue(FilterType.None)]
```
Comment (sharwell) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: :exclamation: I'm not convinced this is correct. I would prefer to see it moved to its own pull request and evaluated separately.

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs [hunk 19]
```diff
@@ -280,6 +288,7 @@ public string[] SelectedIds
             }
         }
 
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2002:DoNotLockOnObjectsWithWeakIdentity")]
```
Comment (sharwell) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: :bulb: I'm not convinced this is correct. I would prefer to see it moved to its own pull request and evaluated separately.

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs [hunk 20]
```diff
@@ -348,6 +357,7 @@ public void Add(string aId, string[] aParentIds, DataType aType, GitRevision aDa
             UpdateData();
         }
 
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2002:DoNotLockOnObjectsWithWeakIdentity")]
```
Comment (sharwell) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: :bulb: I'm not convinced this is correct. I would prefer to see it moved to its own pull request and evaluated separately.

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs [hunk 21]
```diff
@@ -582,6 +594,7 @@ private void BackgroundThreadEntry()
             }
         }
 
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2002:DoNotLockOnObjectsWithWeakIdentity")]
```
Comment (sharwell) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: :bulb: I'm not convinced this is correct. I would prefer to see it moved to its own pull request and evaluated separately.

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs [hunk 22]
```diff
@@ -546,6 +557,7 @@ private void dataGrid_Scroll(object sender, ScrollEventArgs e)
             UpdateColumnWidth();
         }
 
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2002:DoNotLockOnObjectsWithWeakIdentity")]
```
Comment (sharwell) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: :bulb: I'm not convinced this is correct. I would prefer to see it moved to its own pull request and evaluated separately.

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs [hunk 23]
```diff
@@ -462,6 +472,7 @@ private void RebuildGraph()
             Invalidate(true);
         }
 
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2002:DoNotLockOnObjectsWithWeakIdentity")]
```
Comment (sharwell) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: :bulb: I'm not convinced this is correct. I would prefer to see it moved to its own pull request and evaluated separately.

File: GitUI/SpellChecker/EditNetSpell.cs [hunk 24]
```diff
@@ -979,5 +979,25 @@ private void AutoCompleteToolTipTimer_Tick (object sender, EventArgs e)
...
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _autoCompleteListTask.Dispose();
+                _autoCompleteCancellationTokenSource.Dispose();
+                _customUnderlines.Dispose();
+            }
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/SpellChecker/EditNetSpell.cs]: :bulb: This condition could be combined with the previous block.

File: GitUI/Editor/FileViewer.cs [hunk 25]
```diff
@@ -984,5 +984,22 @@ private void copyOldVersionToolStripMenuItem_Click(object sender, EventArgs e)
             CopyNotStartingWith('+');
         }
 
+        /// <summary> 
+        /// Clean up any resources being used.
+        /// </summary>
+        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
+        protected override void Dispose(bool disposing)
+        {
+            GitUICommandsSourceSet -= FileViewer_GitUICommandsSourceSet;
```
Comment (sharwell) [inline GitUI/Editor/FileViewer.cs]: :exclamation: This should only occur when `disposing` is true.

File: GitUI/Editor/FileViewer.cs [hunk 26]
```diff
@@ -984,5 +984,22 @@ private void copyOldVersionToolStripMenuItem_Click(object sender, EventArgs e)
...
+
+            if (disposing)
+                _async.Dispose();
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/Editor/FileViewer.cs]: :bulb: this could be combined with the previous block.

File: GitUI/FilterBranchHelper.cs [hunk 27]
```diff
@@ -165,5 +165,25 @@ private void toolStripBranches_Leave(object sender, EventArgs e)
...
             ApplyBranchFilter(true);
         }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~FilterBranchHelper()
```
Comment (sharwell) [inline GitUI/FilterBranchHelper.cs]: :exclamation: This finalizer is unnecessary and should be removed.

File: GitUI/FilterRevisionsHelper.cs [hunk 28]
```diff
@@ -161,5 +161,28 @@ private void diffContainsToolStripMenuItem_Click(object sender, EventArgs e)
...
             _NO_TRANSLATE_RevisionGrid.SetLimit(limit);
         }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~FilterRevisionsHelper()
```
Comment (sharwell) [inline GitUI/FilterRevisionsHelper.cs]: :exclamation: This finalizer is unnecessary and should be removed.

File: GitUI/Editor/FindAndReplaceForm.cs [hunk 29]
```diff
@@ -318,6 +318,22 @@ internal void SetFileLoader(Func<bool, Tuple<int, string>> fileLoader)
...
+        /// <summary>
+        /// Clean up any resources being used.
+        /// </summary>
+        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+                _search.Dispose();
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/Editor/FindAndReplaceForm.cs]: :bulb: consider combining with previous block

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 30]
```diff
@@ -3218,5 +3222,31 @@ private void FormBrowse_Activated(object sender, EventArgs e)
...
+                _pushButton.Dispose();
+                _pullButton.Dispose();
+#endif
+                _submodulesStatusImagesCTS.Dispose();
+                _formBrowseMenus.Dispose();
+                _filterRevisionsHelper.Dispose();
+                _filterBranchHelper.Dispose();
+            }
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormBrowse.cs]: :bulb: consider combining with previous block

File: GitUI/CommandsDialogs/BrowseDialog/FormBrowseMenus.cs [hunk 31]
```diff
@@ -216,6 +216,26 @@ private IEnumerable<MenuCommand> GetNavigateAndViewMenuCommands()
...
             }
         }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~FormBrowseMenus()
```
Comment (sharwell) [inline GitUI/CommandsDialogs/BrowseDialog/FormBrowseMenus.cs]: :exclamation: This finalizer is unnecessary and should be removed.

File: GitUI/CommandsDialogs/FormClone.cs [hunk 32]
```diff
@@ -304,5 +304,23 @@ private void Branches_DropDown(object sender, EventArgs e)
...
             LoadBranches();
         }
+
+        /// <summary>
+        /// Clean up any resources being used.
+        /// </summary>
+        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
+        protected override void Dispose(bool disposing)
+        {
+            _branchListLoader.Cancel();
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormClone.cs]: :exclamation: only when `disposing` is true

File: GitUI/CommandsDialogs/FormClone.cs [hunk 33]
```diff
@@ -304,5 +304,23 @@ private void Branches_DropDown(object sender, EventArgs e)
...
+
+            if (disposing)
+                _branchListLoader.Dispose();
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormClone.cs]: :bulb: consider combining with previous block

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 34]
```diff
@@ -2436,6 +2435,32 @@ private void toolAuthor_Leave(object sender, EventArgs e)
...
+        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _unstagedLoader.Cancel();
+                _unstagedLoader.Dispose();
+            }
+
+            if (disposing && _interactiveAddBashCloseWaitCts != null)
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormCommit.cs]: :bulb: consider combining with previous block

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 35]
```diff
@@ -2436,6 +2435,32 @@ private void toolAuthor_Leave(object sender, EventArgs e)
...
+            {
+                _interactiveAddBashCloseWaitCts.Cancel();
+                _interactiveAddBashCloseWaitCts.Dispose();
+                _interactiveAddBashCloseWaitCts = null;
+            }
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormCommit.cs]: :bulb: consider combining with previous block(s)

File: GitUI/CommandsDialogs/FormFileHistory.cs [hunk 36]
```diff
@@ -424,5 +424,26 @@ private void Blame_CommandClick(object sender, CommitInfo.CommandEventArgs e)
...
+        {
+            if (disposing)
+            {
+                _asyncLoader.Cancel();
+                _asyncLoader.Dispose();
+                _filterRevisionsHelper.Dispose();
+                _filterBranchHelper.Dispose();
+            }
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormFileHistory.cs]: :bulb: consider combining with previous block

File: GitUI/CommandsDialogs/BrowseDialog/FormGoToCommit.cs [hunk 37]
```diff
@@ -231,5 +231,26 @@ private void SetCommitExpressionFromClipboard()
...
+        {
+            if (disposing)
+            {
+                _tagsLoader.Cancel();
+                _tagsLoader.Dispose();
+                _branchesLoader.Cancel();
+                _branchesLoader.Dispose();
+            }
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/CommandsDialogs/BrowseDialog/FormGoToCommit.cs]: :bulb: consider combining with previous block

File: GitUI/CommandsDialogs/FormPush.cs [hunk 38]
```diff
@@ -845,5 +845,24 @@ private void _NO_TRANSLATE_Branch_SelectedIndexChanged(object sender, EventArgs
...
+        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _remoteBranchesLoader.Cancel();
+                _remoteBranchesLoader.Dispose();
+            }
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormPush.cs]: :bulb: consider combining with previous block

File: GitUI/CommandsDialogs/FormVerify.cs [hunk 39]
```diff
@@ -275,5 +275,24 @@ private GitRevision GetCurrentGitRevision()
...
+        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                selectedItemsHeader.Detach();
+                selectedItemsHeader.Dispose();
+            }
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormVerify.cs]: :bulb: consider combining with previous block

File: GitUI/GitModuleControl.cs [hunk 40]
```diff
@@ -9,12 +9,14 @@ namespace GitUI
     /// <see cref="GitModule"/> and <see cref="GitUICommands"/>.</summary>
     public class GitModuleControl : GitExtensionsControl
     {
+        private object _lock = new object();
```
Comment (sharwell) [inline GitUI/GitModuleControl.cs]: :exclamation: add `readonly`

File: GitUI/GitUICommands.cs [hunk 41]
```diff
@@ -1067,6 +1067,7 @@ public bool StartCherryPickDialog(IWin32Window owner, GitRevision revision)
             return DoActionOnRepo(owner, true, true, PreCherryPick, PostCherryPick, action);
         }
 
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2202:Do not dispose objects multiple times")]
```
Comment (sharwell) [inline GitUI/GitUICommands.cs]: :bulb: in hope that there is a better approach, I would prefer to see this moved to its own (separate) pull request.

File: GitUI/Editor/RichTextBoxXhtmlSupportExtension.cs [hunk 42]
```diff
@@ -33,7 +33,7 @@ static class RichTextBoxXhtmlSupportExtension
         private static int BeginUpdate(HandleRef handleRef)
         {
             // Prevent the control from raising any events.
-            int oldEventMask = NativeMethods.SendMessage(handleRef,
+            int oldEventMask = (int) NativeMethods.SendMessage(handleRef,
```
Comment (sharwell) [inline GitUI/Editor/RichTextBoxXhtmlSupportExtension.cs]: :question: Doesn't it defeat the purpose of the `IntPtr` if you just cast it to an `int` anyway?

File: GitUI/SpellChecker/SpellCheckEditControl.cs [hunk 43]
```diff
@@ -197,5 +198,26 @@ private enum DrawType
...
 
         #endregion
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~SpellCheckEditControl()
```
Comment (sharwell) [inline GitUI/SpellChecker/SpellCheckEditControl.cs]: :exclamation: This finalizer is unnecessary and should be removed.

File: GitUI/CommandsDialogs/RepoHosting/ViewPullRequestsForm.cs [hunk 44]
```diff
@@ -426,5 +426,24 @@ private void _refreshCommentsBtn_Click(object sender, EventArgs e)
...
+        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                loader.Cancel();
+                loader.Dispose();
+            }
+
+            if (disposing && (components != null))
```
Comment (sharwell) [inline GitUI/CommandsDialogs/RepoHosting/ViewPullRequestsForm.cs]: :bulb: consider combining with previous

File: GitUI/UserControls/ToolStripGitStatus.cs [hunk 45]
```diff
@@ -103,10 +103,11 @@ public ToolStripGitStatus()
             _gitDirWatcher.NotifyFilter = NotifyFilters.FileName | NotifyFilters.LastWrite;            
         }
 
-        private void GitUICommandsChanged(IGitUICommandsSource source, GitUICommands oldCommands)
+        private void GitUICommandsChanged(object sender, GitUICommandsChangedEventArgs e)
         {
-            if (oldCommands != null)
+            if (e != null)
```
Comment (sharwell) [inline GitUI/UserControls/ToolStripGitStatus.cs]: :exclamation: `e` should never be null. Pass a non-null instance of the `EventArgs` class even if the property it contains is null.

**Comment by sharwell** [comment]
I finished my review :smile: In general, I'll let the maintainers and/or author determine the validity and course of action for any feedback. I'm subscribed to the events for this in case you have specific questions or comments.

**Comment by mnaoumov** [comment]
Thanks @sharwell , I will apply the most of your suggestions. However, some of your suggestions won't fit in this pull request as its purpose is to fix the warnings only. In some cases the issues you are referring to are real issues but they existed even before I touched the code so it would be a bit illogical to fix it in this scope. So I will create a separate pull request to cover those.

**Comment by sharwell** [comment]
@mnaoumov That makes complete sense (to me at least). Very often I will create an issue that essentially says "XXXX was discovered during review of pull request #2738 and should be addressed. See comments in that pull request for more information." :smile:

**Comment by sharwell** [comment]
@mnaoumov I updated the emoji from my comments. I used :exclamation: to mark items which I believe should be corrected as part of this pull request. I used :grey_exclamation: to mark items which I believe should be corrected, either as part of this pull request or by creating a specific issue in the issue tracker to address them later.

File: GitUI/BuildServerIntegration/BuildServerWatcher.cs (hunk 13 above)
Comment (mnaoumov) [inline GitUI/BuildServerIntegration/BuildServerWatcher.cs]: @sharwell Technically you are right and MemoryStream disposed twice. How do you suggest to fix it? If we omit using on either MemoryStream or StreamReader it would be even more akward.

And here in the comments Jon Skeet agrees with me http://stackoverflow.com/questions/1065168/does-disposing-streamreader-close-the-stream

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs (hunk 18 above)
Comment (mnaoumov) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: I decided to take an opposite approach. I keep my suppress warning attribute and created a separate issue to review that and fix properly #2748

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs (hunk 19 above)
Comment (mnaoumov) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: See #2748

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs (hunk 20 above)
Comment (mnaoumov) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: See #2748

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs (hunk 21 above)
Comment (mnaoumov) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: See #2748

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs (hunk 22 above)
Comment (mnaoumov) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: See #2748

File: GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs (hunk 23 above)
Comment (mnaoumov) [inline GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs]: See #2748

File: GitUI/GitUICommands.cs (hunk 41 above)
Comment (mnaoumov) [inline GitUI/GitUICommands.cs]: See #2749

File: GitCommands/Settings/SettingsCache.cs [hunk 46]
```diff
@@ -16,19 +16,14 @@ public SettingsCache()
 
         public void Dispose()
         {
-            DisposeImpl();
+            Dispose(true);
             GC.SuppressFinalize(this);
         }
 
-        protected virtual void DisposeImpl()
+        protected virtual void Dispose(bool disposing)
```
Comment (jbialobr) [inline GitCommands/Settings/SettingsCache.cs]: What is the purpose of the `disposing` argument? It is always being invoked with the `true` value.

Comment (mnaoumov) [inline GitCommands/Settings/SettingsCache.cs]: It is a well known Dispose pattern - https://msdn.microsoft.com/en-us/library/b1yfkh5e%28v=vs.110%29.aspx

File: GitUI/BuildServerIntegration/BuildServerWatcher.cs (hunk 13 above)
Comment (sharwell) [inline GitUI/BuildServerIntegration/BuildServerWatcher.cs]: @mnaoumov One option is adding `Justification = "http://stackoverflow.com/questions/1065168/does-disposing-streamreader-close-the-stream"` to the `SuppressMessage` attribute.

**Comment by KindDragon** [comment]
Merged
